hop/io.py

```diff
@@ -509,13 +919,106 @@ def write_raw(self, packed_message, headers=None,
-                             topic=topic)
+        Args:
+            message: The message to be sent which must already be correctly encoded by
+                     :meth:`Producer.pack <hop.io.Producer.pack>`
+            headers: Any headers to attach to the message as a list of (str, bytes) 2-tuples.
+        Return: A tuple consisting of the replacement message payload, replacement message headers,
+                and an error object. In the case of success the first two tuple entries are valid
+                and suitable for passing to the Kafka producer's write, and the third tuple entry
+                is None. In the case of failure, the first two entries are None, and the error is a
+                confluent_kafka.KafkaError object.
+        """
+        offload_suffix = "+oversized"
+        write_url = f"{self.offload_url}/topic/{topic}{offload_suffix}"
+        msg_id = None
+        test = False
+        for header in headers:
+            if header[0] == "_test" and header[1] == b"true":
+                test = True
+            if header[0] == "_id":
+                try:
+                    msg_id = str(uuid.UUID(bytes=header[1]))
+                except ValueError:
+                    pass
+        if msg_id is None:
+            err = confluent_kafka.KafkaError(confluent_kafka.KafkaError._BAD_MSG,
+                                             "Message has no ID ('_id' header). "
+                                             "Was it properly processed by Producer.pack?",
+                                             False, False, False)
+            return (None, None, err)
+        data = bson.dumps({"message": message, "headers": headers})
+        try:
+            # We assume that no server will allow un-authenticated writes, so we use shortcut=True
+            # to start attempting a SCRAM handshake as quickly as possible.
+            resp = requests.post(write_url, data=data,
+                                 auth=http_scram.SCRAMAuth(self.auth, shortcut=True))
+        except RuntimeError as ex:
+            err = confluent_kafka.KafkaError(confluent_kafka.KafkaError.SASL_AUTHENTICATION_FAILED,
+                                             "Failed to send large message to offload server at "
+                                             f"{write_url}: {str(ex)}", False, False, False)
+            return (None, None, err)
+        if not resp.ok:
+            err = _http_error_to_kafka(resp.status_code,
+                                       "Failed to send large message to offload server at "
+                                       f"{write_url}: POST request failed with status "
+                                       f"{resp.status_code}: {resp.content}")
+            return (None, None, err)
+        msg_url = f"{self.offload_url}/msg/{msg_id}"
+        placeholder = models.ExternalMessage(url=msg_url)
+        # It is important that the placeholder be assigned its own ID, so it does not collide in the
```

Comment:
Would it make sense to have this id be the original message id + some fixed known string, so you can translate between the two easily? I.e. 1234 -> 1234-p for the placeholder?

Reply:
We've currently defined our IDs as UUIDv4, so we don't have a valid way to step out of the 'standard ID namespace' and still have a valid ID. I don't think anything in our design actually depends using UUIDv4, so we could change it, but we would have to plumb that through the archive implementation as well as changing the docs. Also, at the moment it is allowed, (although obviously no one is yet doing it) to have multiple placeholders refer to the same original data. 

So, we could do this, but I'm not sure what we would gain would be worth it.